### PR TITLE
Visitors can create an account and log in.

### PR DESCRIPTION
**Pull request by athifw**

### CODE REVIEW

File: app/assets/javascripts/logins.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (efatsi) [inline app/assets/javascripts/logins.coffee]: :fire:

File: app/assets/javascripts/users.coffee [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (efatsi) [inline app/assets/javascripts/users.coffee]: :fire:

File: app/assets/stylesheets/logins.scss [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Logins controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (efatsi) [inline app/assets/stylesheets/logins.scss]: :fire:

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -2,4 +2,25 @@ class ApplicationController < ActionController::Base
...
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
+
+  private
+  def current_user
+     @current_user ||= begin
+      id = session[:current_user_id]
+      
+      if id
+        User.find_by(id: id)
```
Comment (efatsi) [inline app/controllers/application_controller.rb]: `User.find(id)` is shorthand for this.

File: app/helpers/logins_helper.rb [hunk 5]
```diff
@@ -0,0 +1,2 @@
+module LoginsHelper
+end
```
Comment (efatsi) [inline app/helpers/logins_helper.rb]: more :fire:

File: app/helpers/users_helper.rb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+module UsersHelper
+end
```
Comment (efatsi) [inline app/helpers/users_helper.rb]: these helpers are auto-generated and rarely used. if you ever do need a helper method for your views, ApplicationHelper is a decent place for them, but considering the practice of using them is often eclipsed in goodness by other options, a typical app will only have a handful - negating the necessity for individual helpers.

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
+class SessionsController < ApplicationController
+  
+  def new
+
+  end
+
+  def create
+    
+    user = User.authenticate(params[:email], params[:password])
```
Comment (efatsi) [inline app/controllers/sessions_controller.rb:7]: does this always return `nil` if the email/password passed in is invalid?

File: app/controllers/application_controller.rb (hunk 4 above)
Comment (zporter) [inline app/controllers/application_controller.rb]: @efatsi -- Doesn't `find_by` return `nil` and `find` raise an exception if the record cannot be found?

Comment (efatsi) [inline app/controllers/application_controller.rb]: correct. shame upon me.

File: app/controllers/sessions_controller.rb (hunk 7 above)
Comment (athifw) [inline app/controllers/sessions_controller.rb:7]: it does! will add a test soon

**Comment by efatsi** [comment]
this is very nitpicky, but you have 3 unnecessary new lines in `sessions/new.html.erb`, as well as one each in the `new` and `create` actions of `sessions_controller.rb`, and at the end of your user specs.

File: app/controllers/application_controller.rb (hunk 4 above)
Comment (athifw) [inline app/controllers/application_controller.rb]: I talked with Chris about this and we think it's a better idea to raise an exception if there's a weird id in the session. So I'll change this too

File: app/views/users/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  <br>
+  <%= f.submit %>
+<% end %>
+
+<% if @user.errors.any? %>
+  <% @user.errors.full_messages.each do |error| %>
+    <%= error %>
+    <br>
+  <% end %>
+<% end %>
```
Comment (zporter) [inline app/views/users/new.html.erb:22]: Generally, I prefer the errors to render before the form so I immediately see that something went wrong with my submission and I need to check my inputs. That's just one man's opinion.

File: spec/features/sessions_spec.rb [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+      it "display errors" do        
+        click_button 'Log in'
+        expect(page).to have_content("Invalid email or password.")
+      end
+    end
+    context "with valid credentials" do
+      let!(:user) { create :user, email: 'bob@test.com', username: 'Bob' }
+      it "logs in successfully" do
+        fill_in 'Email', with: 'bob@test.com'
+        fill_in 'Password', with: 'password'
```
Comment (zporter) [inline spec/features/sessions_spec.rb:19]: How do we know the user's password is 'password'?  Is that in the factory definition?

File: spec/models/user_spec.rb [hunk 10]
```diff
@@ -0,0 +1,33 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  subject { create :user }
+  
+  it { should validate_presence_of :username }
+  it { should validate_presence_of :email }
+  it { should validate_uniqueness_of :username }
+  it { should validate_uniqueness_of :email }
+
```
Comment (zporter) [inline spec/models/user_spec.rb:11]: check out adding [`it { should have_secure_password }`](http://matchers.shoulda.io/docs/v2.8.0/Shoulda/Matchers/ActiveModel.html#have_secure_password-instance_method)

File: spec/models/user_spec.rb [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+  it { should validate_uniqueness_of :email }
+  
+  describe "to_s" do
+    it "returns username" do
+      subject.username = "Test name"
+      expect(subject.to_s).to eq("Test name")
+    end
+  end
+
+  describe "authenticate" do
```
Comment (efatsi) [inline spec/models/user_spec.rb]: It's conventional to add a `.` before class methods, and a `#` before instance methods in specs. Gives the reader of the specs more instant context into what's being tested.

**Comment by zporter** [comment]
Dropped a few comments and +1 to @efatsi's nitpicks. Otherwise, lookin' good :+1:

**Comment by efatsi** [comment]
Great work man :+1:

File: spec/features/sessions_spec.rb (hunk 9 above)
Comment (athifw) [inline spec/features/sessions_spec.rb:19]: Yeah it's in factory.

Comment (zporter) [inline spec/features/sessions_spec.rb:19]: Probably better to specify in the `create` call above or just reference `user.password` here.  That way, if the password in the factory ever changed, you wouldn't need to update it here.

**Pull request merged**